tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {backend_util} from '@tensorflow/tfjs-core';
+
+import {getCoordsDataType} from '../shader_preprocessor';
+import {computeDispatch} from '../webgpu_util';
+
+import {WebGPUProgram} from './webgpu_program';
+
+export class BatchNormProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  userCode: string;
+  dispatchLayout: {x: number[], y: number[], z: number[]};
+  dispatch: [number, number, number];
+  variableNames: string[];
+  workGroupSize: [4, 4, 4];
+
+  constructor(
+      xShape: number[], meanShape: number[], varianceShape: number[],
+      offsetShape: number[]|null, scaleShape: number[]|null,
+      varianceEpsilon: number) {
+    this.variableNames = ['x', 'mean', 'variance'];
+    backend_util.assertAndGetBroadcastShape(xShape, meanShape);
+    backend_util.assertAndGetBroadcastShape(xShape, varianceShape);
+    this.outputShape = xShape;
+    this.dispatchLayout = {x: [1, 2], y: [0], z: [3]};
+    const dim = this.outputShape.length;
+    const coordsDataType = getCoordsDataType(dim);
+    let setOutput =
+        'setOutput(coords[0], coords[1], coords[2], coords[3], value);';
+    switch (dim) {
+      case 2:
+        this.dispatchLayout = {x: [1], y: [0], z: []};
+        setOutput = 'setOutput(coords[0], coords[1], value);';
+        break;
+      case 3:
+        this.dispatchLayout = {x: [1, 2], y: [0], z: []};
+        setOutput = 'setOutput(coords[0], coords[1], coords[2], value);';
+        break;
+      case 4:
+        this.dispatchLayout = {x: [1, 2], y: [0], z: [3]};
+        setOutput =
+            'setOutput(coords[0], coords[1], coords[2], coords[3],value);';
+        break;
+      default:
+        this.dispatchLayout = {x: [1, 2], y: [0], z: [3]};
+        setOutput =
+            'setOutput(coords[0], coords[1], coords[2], coords[3],value);';
```

Comment:
Can we use if/else rather than switch? Also we know that the output and x are 4D - can we avoid this logic altogether? We often reshape tensors before passing them to kernels to make it possible to simplify shader code like this.

Reply:
I fix the syntax. I think this logic needs to be preserved. The x and output may be 2D, 3D, or 4D because ops is modified in #2996 ; batchnorm_2d, batchnorm_3d, and batchnorm_4d will call this kernel.